pkg/hooks/manager.go

```diff
@@ -260,75 +279,152 @@ func (m *hookManager) onDefineDomainCallback(callback *callBackClient, domainSpe
+			log.Log.Reason(err).Error("Failed to unmarshal CloudInitData result")
+			return nil, err
+		}
+
+		if !cloudinit.IsValidCloudInitData(resultData) {
+			// Be backwards compatible for hook sidecars still working on CloudInitNoCloudSource objects instead of CloudInitData
+			var resultNoCloudSourceData *v1.CloudInitNoCloudSource
+			err = json.Unmarshal(noCloudSource, &resultNoCloudSourceData)
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to unmarshal CloudInitNoCloudSource result")
+				return nil, err
 			}
+			resultData = &cloudinit.CloudInitData{
+				DataSource:  cloudInitData.DataSource,
+				UserData:    resultNoCloudSourceData.UserData,
+				NetworkData: resultNoCloudSourceData.NetworkData,
+			}
+		}
+		return resultData, nil
+	}
+
+	callbacks, found := m.CallbacksPerHookPoint[hooksInfo.PreCloudInitIsoHookPointName]
+	if !found {
+		return cloudInitData, nil
+	}
+
+	for _, callback := range callbacks {
+		switch callback.Version {
+		case hooksV1alpha2.Version:
+			json, err := dataToJSON()
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to run PreCloudInitIso")
+				return cloudInitData, err
+			}
+
+			conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
+			if err != nil {
+				log.Log.Reason(err).Errorf(dialSockErr, callback.SocketPath)
+				return cloudInitData, err
+			}
+			defer conn.Close()
+
+			client := hooksV1alpha2.NewCallbacksClient(conn)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			result, err := client.PreCloudInitIso(ctx, &hooksV1alpha2.PreCloudInitIsoParams{
+				CloudInitData:          json[0],
+				CloudInitNoCloudSource: json[1],
+				Vmi:                    json[2],
+			})
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to call PreCloudInitIso")
+				return cloudInitData, err
+			}
+			return validateResult(result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
+		case hooksV1alpha3.Version:
+			json, err := dataToJSON()
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to run PreCloudInitIso")
+				return cloudInitData, err
+			}
+
+			conn, err := grpcutil.DialSocketWithTimeout(callback.SocketPath, 1)
+			if err != nil {
+				log.Log.Reason(err).Errorf(dialSockErr, callback.SocketPath)
+				return cloudInitData, err
+			}
+			defer conn.Close()
+
+			client := hooksV1alpha3.NewCallbacksClient(conn)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+			defer cancel()
+
+			result, err := client.PreCloudInitIso(ctx, &hooksV1alpha3.PreCloudInitIsoParams{
+				CloudInitData:          json[0],
+				CloudInitNoCloudSource: json[1],
+				Vmi:                    json[2],
+			})
+			if err != nil {
+				log.Log.Reason(err).Error("Failed to call PreCloudInitIso")
+				return cloudInitData, err
+			}
+			return validateResult(result.GetCloudInitData(), result.GetCloudInitNoCloudSource())
+		default:
+			log.Log.Errorf("Unsupported callback version: %s", callback.Version)
```

Comment:
In conjunction with the above comment, is there a reason for changing this as well, rather than the original code that panicked? The statement here should still hold true.
```
panic("Should never happen, version compatibility check is done during Info call")
```

Reply:
Yes, we should avoid panic. I see this as just another opportunity to avoid `virt-launcher` to panic due bad sidecar.